src/gf.c

```diff
@@ -3071,6 +3077,129 @@ JL_DLLEXPORT jl_value_t *jl_apply_generic(jl_value_t *F, jl_value_t **args, uint
+
+    // To get the true dynamic types tuple, we need to fetch the typeof from any boxes.
+    for (int i = 1; i < ntypes; ++i) {
+        //jl_printf(JL_STDERR, "\ntypes[%d]: ", i);
+        //jl_(types[i]);
+        //jl_printf(JL_STDERR, "\n is concrete: %d\n", jl_is_concrete_type(types[i]));
+        if (!jl_is_concrete_type(types[i])) {
+            //jl_(args[i-1]);
+            // Get the runtime type
+            types[i] = jl_typeof(args[i-1]);
+        }
+        //jl_(types[i]);
+    }
+
+    jl_datatype_t* tt = (jl_datatype_t*)jl_apply_tuple_type_v(types, ntypes);
+    //jl_(tt);
+
+    //jl_method_match_t *match = _gf_invoke_lookup((jl_value_t*)tt, jl_nothing, world, &min_valid, &max_valid);
+    ////jl_printf(JL_STDERR, "match: %p\n", match);
+    //if (match == NULL) {
+    //    //jl_printf(JL_STDERR, "No match\n");
+    //    return jl_nothing;
+    //}
+    //jl_method_instance_t *mfunc = jl_method_match_to_mi(match, world, world, world, 1);
+
+
+    // TODO: Share the method caching from jl_apply_generic.
+    // cache miss case
+    JL_TIMING(METHOD_LOOKUP_SLOW, METHOD_LOOKUP_SLOW);
+    jl_methtable_t *mt = jl_gf_mtable(F);
+    jl_method_instance_t *mfunc = jl_mt_assoc_by_type(mt, tt, world);
+
+    if (mfunc == NULL) {
+#ifdef JL_TRACE
+        if (error_en)
+            show_call(F, args, nargs);
+#endif
+        jl_method_error(F, args, nargs, world);
+        // unreachable
+    }
+
+
+    //jl_printf(JL_STDERR, "mfunc: %p\n", mfunc);
+    //jl_printf(JL_STDERR, "spectypes: "); jl_(mfunc->specTypes);
+    // jl_method_instance_t *mfunc = jl_lookup_generic_(F, args, nargs,
+    //                                                  jl_int32hash_fast(jl_return_address()),
+    //                                                  world);
+
+    //jl_value_t** newargs = (jl_value_t**)alloca(sizeof(jl_value_t*) * nargs);
+    //jl_printf(JL_STDERR, "new args:\n");
+
+    // Matched specialized types
+    jl_svec_t *spec_types = ((jl_datatype_t*)mfunc->specTypes)->parameters;
+
+    // Reuse the array of args, but point to the newly boxed args when reboxed.
+    jl_value_t** newargs = (jl_value_t**)args;
+
+    //jl_datatype_t* tt = (jl_datatype_t*)types;
+    for (size_t i = 0; i < nargs; i++) {
+        jl_datatype_t* typ = jl_svecref(tt->parameters, i+1); // skip the function
+        //jl_printf(JL_STDERR, "Expecting type %zu: ", i);
+        //jl_(jl_svecref(spec_types, i+1));
+        // If the function is expecting a boxed value (because it's not specialized)
+        // we need to ensure the value is boxed.
+        if (!jl_is_concrete_type(jl_svecref(spec_types, i+1)) &&
+            // Does this need to be "and isbits"? I don't _think_ so?
+            jl_is_concrete_type(jl_typeof(args[i]))) {
+            // re-box the value
+            jl_printf(JL_STDERR, "reboxing arg %zu\n", i);
+            newargs[i] = rebox_type_and_bytes((jl_datatype_t*)typ, args[i]);
+        } else {
+            newargs[i] = args[i];
```

Comment:
You are implicitly casting the void* to a jl_value_t* here (that cast is explicitly permitted in C without warning), which is why this approach won't work.

Reply:
Ah, but we're only doing that IF the pointee **is** a real boxed object. Which we know because the `if` above exactly matches the same `if` we use in codegen:

i.e. we are only taking the address of stack values _iff_ the object isn't already boxed. If it _is_ already boxed, we just pass the box through. And that's what we are checking here!

--------

To make this more explicit, we could separate the arguments into two different lists, one for the valid julia objects (jl_value_t**) and one for the stack objects (void**)? The reason i didn't do that is because then you'd also need a third list to indicate the interleaving order of the arguments.

Does that make sense and answer your objection?